src/tactic/field_simp.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2019 Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+-/
+
+import tactic.interactive
+import tactic.norm_num
+
+/-!
+# `field_simp` tactic
+
+Tactic to clear denominators in algebraic expressions, based on `simp` with a specific simpset.
+-/
+
+namespace tactic
+
+/-- Try to prove a goal of the form `x ≠ 0` by calling `assumption`, or `norm_num1` if `x` is
+a numeral. -/
+meta def field_simp.ne_zero : tactic unit := do
+  goal ← tactic.target,
+  match goal with
+  | `(%%e ≠ 0) := assumption <|> do n ← e.to_rat, `[norm_num1]
+  | _ := tactic.fail "goal should be of the form `x ≠ 0`"
+  end
+
+namespace interactive
+open interactive interactive.types
+
+/--
+The goal of `field_simp` is to reduce an expression in a field to an expression of the form `n / d`
+where neither `n` nor `d` contains any division symbol, just using the simplifier (with a carefully
+crafted simpset named `field_simps`) to reduce the number of division symbols whenever possible by
+iterating the following steps:
+
+- write an inverse as a division
+- in any product, move the division to the right
+- if there are several divisions in a product, group them together at the end and write them as a
+  single division
+- reduce a sum to a common denominator
+
+If the goal is an equality, this simpset will also clear the denominators, so that the proof
+can normally be concluded by an application of `ring` or `ring_exp`.
+
+`field_simp [hx, hy]` is a short form for
+`simp [-one_div, -mul_eq_zero, hx, hy] with field_simps {discharger := [field_simp.ne_zero]}`
+
+Note that this naive algorithm will not try to detect common factors in denominators to reduce the
+complexity of the resulting expression. Instead, it relies on the ability of `ring` to handle
+complicated expressions in the next step.
+
+As always with the simplifier, reduction steps will only be applied if the preconditions of the
+lemmas can be checked. This means that proofs that denominators are nonzero should be included. The
+fact that a product is nonzero when all factors are, and that a power of a nonzero number is
+nonzero, are included in the simpset, but more complicated assertions (especially dealing with sums)
```

Comment:
Since you've changed the discharger from the default `simp`, this claim about products and powers is no longer true, right? The only preconditions you'll solve are those that are literally in the context and numeric ones.

Reply:
I don't get it. The default discharger for `simp` is `none`, right? So it's only making things more powerful. Indeed, with the modification it solves everything it solved previously (at least all the mathlib examples and tests).

Reply:
Ah, sorry, I take it back. Seems like the `discharger` option is used in addition to recursively calling `simp`. I thought it was replacing that.